tools/windows/DatadogInterop/DatadogInterop/msstoreapps.cpp

```diff
@@ -0,0 +1,123 @@
+    // Convert ticks to milliseconds
+    int64_t millisSince1601 = ticks / 10000; // 10,000 * 100ns = 1ms
+    int64_t unixMillis = millisSince1601 - EPOCH_DIFF_MILLIS;
+    return unixMillis;
+}
+
+static const wchar_t *copyHStr(MSStoreInternal *msStore, winrt::hstring hstr) {
+    // Add to msStore's string vector to keep hstring alive
+    msStore->strings.push_back(hstr);
+    return hstr.c_str();
+}
+
+static void addEntryToStore(MSStoreInternal *msStore, const Package &pkg, winrt::hstring displayName) {
+    auto id = pkg.Id();
+    int64_t installDate = 0;
+    // Not all packages have InstalledDate
+    try {
+        installDate = dtToUnixEpochMs(pkg.InstalledDate());
+    } catch (...) {
+    }
+    PackageVersion version = id.Version();
+
+    MSStoreEntry e{};
+    e.display_name = copyHStr(msStore, displayName);
+    e.version_major = version.Major;
+    e.version_minor = version.Minor;
+    e.version_build = version.Build;
+    e.version_revision = version.Revision;
+    e.install_date = installDate;
+    e.is_64bit = is64(id.Architecture());
+    e.publisher = copyHStr(msStore, id.Publisher());
+    e.product_code = copyHStr(msStore, id.FamilyName());
+
+    msStore->entriesVec.push_back(e);
+}
+
+extern "C" __declspec(dllexport) int GetStore(MSStore **out) {
+    if (!out) {
+        return Result::INVALID_PARAMS;
+    }
+
+    MSStoreInternal *msStore = new MSStoreInternal();
+    msStore->count = 0;
+    msStore->entries = nullptr;
+
+    try {
+        winrt::init_apartment();
+
+        PackageManager pm;
+
+        auto packages = pm.FindPackagesWithPackageTypes(PackageTypes::Main);
+
+        for (auto const &pkg : packages) {
+            auto id = pkg.Id();
+            auto displayName = id.Name();
+            auto appListEntries = pkg.GetAppListEntries();
+
+            if (appListEntries.Size() == 0) {
+                addEntryToStore(msStore, pkg, displayName);
+            } else {
+                for (auto const &appListEntry : appListEntries) {
+                    auto displayInfo = appListEntry.DisplayInfo();
+                    if (displayInfo) {
+                        auto dn = displayInfo.DisplayName();
+                        if (!dn.empty()) {
+                            displayName = dn;
+                        }
+                    }
+
+                    addEntryToStore(msStore, pkg, displayName);
+                }
+            }
+        }
+
+        msStore->count = static_cast<int32_t>(msStore->entriesVec.size());
+        if (!msStore->entriesVec.empty()) {
+            msStore->entries = msStore->entriesVec.data();
+        }
+
+        *out = msStore;
+        return Result::SUCCESS;
+    } catch (...) {
+        return Result::EXCEPTION;
+    }
+}
+
+extern "C" __declspec(dllexport) int FreeStore(MSStore *msStore) {
+    if (!msStore) {
+        return Result::INVALID_PARAMS;
+    }
+    delete reinterpret_cast<MSStoreInternal *>(msStore);
```

Comment:
you probably better of with dynamic_cast, although I think a single struct would actually sufficient, and additional complexity avoided, if you put two fields accessed by Go in front.

Reply:
dynamic_cast can only be used with polymorphic classes. Also, I still prefer having `MSStore` and `MSStoreInternal` separate because it's not safe exposing C++ data types to CGO. I think separating them is cleaning and it's not that complex.

Reply:
> I still prefer having MSStore and MSStoreInternal separate because it's not safe exposing C++ data types to CGO. 

Up-to you, you are not doing anything wrong. To me personally, that complexity does not justify simplicity of a data model and use cases. Even if you want to access C++ objects embedded into the struct you cannot using regular Go/CGO facility.